src/ripple/overlay/PeerReservationTable.h

```diff
@@ -0,0 +1,142 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_OVERLAY_PEER_RESERVATION_TABLE_H_INCLUDED
+#define RIPPLE_OVERLAY_PEER_RESERVATION_TABLE_H_INCLUDED
+
+#include <ripple/beast/hash/uhash.h>
+#include <ripple/core/DatabaseCon.h>
+#include <ripple/json/json_forwards.h>
+#include <ripple/protocol/PublicKey.h>
+
+#define SOCI_USE_BOOST
+#include <boost/optional.hpp>
+#include <soci/soci.h>
+
+#include <string>
+#include <unordered_map>
+
+namespace ripple {
+
+// Value type for reservations.
+// REVIEWER: What is the preferred naming convention for member variables?
+struct PeerReservation
+{
+public:
+    PublicKey nodeId_;
+    boost::optional<std::string> description_;
+
+    auto
+    toJson() const -> Json::Value;
+};
+
+class PeerReservationTable
+{
+public:
+    using table_type =
+        std::unordered_map<PublicKey, PeerReservation, beast::uhash<>>;
```

Comment:
Every time I've tried to remove the key from a record type that is kept in a map, I've come to regret it. At some point, I want the full record, and then my only choice is a tuple of `(Key, Record)` (speaking abstractly).

Duplicating the key costs 41 bytes on a 64-bit machine, and we don't expect anywhere near 100 peer reservations. I'd rather not slice this record type until it is known that its size is a problem.

Reply:
Your call.  However it's not just about space savings.  It's also about duplication of information, which can lead to bugs.  Certainly your current approach works.  But it also allows for the situation where the `PublicKey` in the key doesn't match the `PublicKey` in the `PeerReservation` value.  Yes, that would be a bug.  But bugs happen.

At any rate, I won't lose sleep over either approach.  Use your best judgement.

Reply:
I realized overnight that, since you want to keep the key in the record type, what you really want is not a `map`.  You want a `set`.  Yes, since you're using a hash table you have to write a special hash function that doesn't include the `description` in the hash.  But, in my opinion, a `set` really matches your design intent.

Again, I won't lose any sleep over the choice you make here.

Reply:
You're exactly right. I'll make that change. Hoping I can understand @HowardHinnant's `hash` enough to write my own overload...

Reply:
If you have any questions, I'm happy to help.  And happy to review.